File: python/get_weather.py
import requests
from collections import defaultdict
from datetime import datetime
import numpy as np
from geojson import get_centroid_latitude_longitude
import time

def calculate_mean_humidity(hourly_data):
    time = hourly_data['time']
    humidity = hourly_data['relative_humidity_2m']
    
    daily_humidity = defaultdict(list)
    
    for i in range(len(time)):
        t = time[i]
        h = humidity[i]
        date = datetime.strptime(t, '%Y-%m-%dT%H:%M').strftime('%Y-%m-%d')
        daily_humidity[date].append(h)
    
    mean_humidity = {}
    for date, humidities in daily_humidity.items():
        if None in humidities:
            mean_humidity[date] = None
        else:
            mean_humidity[date] = np.mean(humidities)
    
    return mean_humidity

def fetch_weather(geometry, start_date, end_date, retries=5):
    latitude, longitude = get_centroid_latitude_longitude(geometry)
    
    url = "https://archive-api.open-meteo.com/v1/archive"
    params = {
        "latitude": latitude,
        "longitude": longitude,
        "start_date": start_date.strftime('%Y-%m-%d'),
        "end_date": end_date.strftime('%Y-%m-%d'),
        "daily": "temperature_2m_mean,precipitation_sum",
        "hourly":"relative_humidity_2m"
    }

    attempt = 0
    while attempt < retries:
        response = requests.get(url, params=params)
        if response.status_code == 200:
            data = response.json()
            data_parsed = {}
            times = data['daily']['time']
            temperatures = data['daily']['temperature_2m_mean']
            precipitations = data['daily']['precipitation_sum']
            humidity = calculate_mean_humidity(data['hourly'])

            for i in range(len(times)):
                date = times[i]
                data_parsed[date] = {
                    "temperature": temperatures[i],
                    "precipitation": precipitations[i],
                    "humidity": humidity[date]
                }
            return data_parsed
        else:
            print(f"Failed to retrieve data: {response.status_code}. Retrying... ({attempt + 1}/{retries})")
            time.sleep(10)
            attempt += 1
    
    print("Failed to retrieve data after multiple attempts.")
    raise Exception("Failed to retrieve data after multiple attempts.")

if __name__ == "__main__":
    fetch_weather(-23.5489, -46.6388, datetime(2024, 11, 29))

File: go-service/internal/weather/get_weather.go
package weather

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"

	"github.com/forest-guardian/forest-guardian-api-poc/internal/cache"
)

type HourlyData struct {
	Time             []string  `json:"time"`
	RelativeHumidity []float64 `json:"relative_humidity_2m"`
}

type DailyData struct {
	Time          []string  `json:"time"`
	Temperature   []float64 `json:"temperature_2m_mean"`
	Precipitation []float64 `json:"precipitation_sum"`
}

type WeatherResponse struct {
	Hourly HourlyData `json:"hourly"`
	Daily  DailyData  `json:"daily"`
}

type Weather struct {
	Precipitation float64
	Temperature   float64
	Humidity      float64
}

type HistoricalWeather map[time.Time]Weather

func calculateMeanHumidity(hourlyData HourlyData) map[string]float64 {
	dailyHumidity := make(map[string][]float64)
	meanHumidity := make(map[string]float64)

	for i, t := range hourlyData.Time {
		h := hourlyData.RelativeHumidity[i]
		date := t[:10] // Extract the date (YYYY-MM-DD)
		dailyHumidity[date] = append(dailyHumidity[date], h)
	}

	for date, humidities := range dailyHumidity {
		var sum float64
		for _, h := range humidities {
			sum += h
		}
		meanHumidity[date] = sum / float64(len(humidities))
	}

	return meanHumidity
}

func FetchWeather(latitude, longitude float64, startDate, endDate time.Time, retries int) (HistoricalWeather, error) {
	weatherCache := cache.NewFileCache[HistoricalWeather]("weather")
	cacheKey := weatherCache.GenerateKey(latitude, longitude, startDate.Format("2006-01-02"), endDate.Format("2006-01-02"))

	// Try to read from cache
	if cached, ok := weatherCache.Get(cacheKey); ok {
		fmt.Printf("Weather cache HIT for key: %s (lat: %.6f, lon: %.6f, %s to %s)\n", 
			cacheKey, latitude, longitude, startDate.Format("2006-01-02"), endDate.Format("2006-01-02"))
		return cached, nil
	}
	
	fmt.Printf("Weather cache MISS for key: %s (lat: %.6f, lon: %.6f, %s to %s)\n", 
		cacheKey, latitude, longitude, startDate.Format("2006-01-02"), endDate.Format("2006-01-02"))

	url := "https://archive-api.open-meteo.com/v1/archive"
	params := fmt.Sprintf("?latitude=%f&longitude=%f&start_date=%s&end_date=%s&daily=temperature_2m_mean,precipitation_sum&hourly=relative_humidity_2m",
		latitude, longitude, startDate.Format("2006-01-02"), endDate.Format("2006-01-02"))

	var weatherData WeatherResponse
	var attempt int

	client := &http.Client{
		Timeout: 30 * time.Second,
	}

	for attempt < retries {
		resp, err := client.Get(url + params)
		if err != nil {
			fmt.Printf("Failed to retrieve data: %v. Retrying... (%d/%d)\n", err, attempt+1, retries)
			time.Sleep(10 * time.Second)
			attempt++
			continue
		}
		defer resp.Body.Close()

		if resp.StatusCode == http.StatusOK {
			fmt.Printf("Response headers: Content-Length=%s, Content-Encoding=%s\n", 
				resp.Header.Get("Content-Length"), resp.Header.Get("Content-Encoding"))
			
			bodyBytes, err := io.ReadAll(resp.Body)
			if err != nil {
				fmt.Printf("Failed to read response body (partial read may have occurred): %v\n", err)
				fmt.Printf("Partial body read (%d bytes): %s\n", len(bodyBytes), string(bodyBytes))
				return nil, fmt.Errorf("failed to read response body: %v", err)
			}
			
			fmt.Printf("Successfully read %d bytes from response body\n", len(bodyBytes))

			err = json.Unmarshal(bodyBytes, &weatherData)
			if err != nil {
				fmt.Printf("Failed to decode JSON response. Body: %s\n", string(bodyBytes))
				return nil, fmt.Errorf("failed to parse response: %v", err)
			}

			// Parse data
			dataParsed := HistoricalWeather{}
			humidity := calculateMeanHumidity(weatherData.Hourly)

			for i, date := range weatherData.Daily.Time {
				parsedDate, err := time.Parse("2006-01-02", date)
				if err != nil {
					return nil, fmt.Errorf("failed to parse date: %v", err)
				}
				dataParsed[parsedDate] = Weather{
					Temperature:   weatherData.Daily.Temperature[i],
					Precipitation: weatherData.Daily.Precipitation[i],
					Humidity:      humidity[date],
				}
			}

			// Write to cache
			if err := weatherCache.Set(cacheKey, dataParsed); err != nil {
				fmt.Printf("Warning: failed to write cache: %v\n", err)
			} else {
				fmt.Printf("Weather cache WRITTEN for key: %s (lat: %.6f, lon: %.6f, %s to %s)\n", 
					cacheKey, latitude, longitude, startDate.Format("2006-01-02"), endDate.Format("2006-01-02"))
			}

			return dataParsed, nil
		} else {
			fmt.Printf("Failed to retrieve data: %d. Retrying... (%d/%d)\n", resp.StatusCode, attempt+1, retries)
			time.Sleep(10 * time.Second)
			attempt++
		}
	}

	return nil, fmt.Errorf("failed to retrieve data after %d attempts", retries)
}
